### DISCUSSION TITLE
Different label categories than expected in a spark-nlp NER model

### DISCUSSION THREAD

**Discussion by ag-din** [JohnSnowLabs/spark-nlp#13167] [Q&A] [answered]
Hello everyone, 

I am a beginner with spark-nlp, and I want to train a NER model that recognises in texts 2 types of entities with SPECSKILL and HUMANSKILL labels. I'm using Python 3.7.12 and spark-nlp 4.2.3. The training and test datasets are in CoNLL 2003 format. I did a first small training and got the following results:
<img width="821" alt="Screenshot 2022-11-28 at 18 44 42" src="https://user-images.githubusercontent.com/21266645/204560318-0fe02e32-c720-4927-b409-f98f821471f5.png">
It is noted that labels=9. However, I expect labels=2 (SPECSKILL and HUMANSKILL only). As it caught my attention, I tried using the conll2003 data [here](https://github.com/JohnSnowLabs/spark-nlp/tree/master/src/test/resources/conll2003), and got similar results with respect to the labels:
<img width="820" alt="Screenshot 2022-11-28 at 18 53 06" src="https://user-images.githubusercontent.com/21266645/204560600-83755114-3521-4ee1-b50a-015d6ebf41cc.png">
It is noted that labels=9. However, I expect labels=4 (LOC, ORG, MISC, and PER only).

**Questions:**
1. Is it normal to get these results in terms of number of labels?
2.  What exactly does the trained NER model recognise that generates other categories than the ones I expect (I expect labels=2, from SPECSKILL & HUMANSKILL)?
3. I am following [CoNLL 2003 documentation, to get the 4th item (entity)](https://www.clips.uantwerpen.be/conll2003/ner/#:%7E:text=Spanish%20and%20Dutch.-,Software%20and%20Data,-The%20CoNLL%2D2003:~:text=The%20chunk%20tags%20and%20the%20named%20entity%20tags%20have%20the%20format%20I%2DTYPE%20which%20means%20that%20the%20word%20is%20inside%20a%20phrase%20of%20type%20TYPE.%20Only%20if%20two%20phrases%20of%20the%20same%20type%20immediately%20follow%20each%20other%2C%20the%20first%20word%20of%20the%20second%20phrase%20will%20have%20tag%20B%2DTYPE%20to%20show%20that%20it%20starts%20a%20new%20phrase), but I see this is causing “harm” (as far as I am concerned), as theoretically equal entities are getting recognized as different (i.e., “B-SPECSKILL” is being considered different to “I-SPECSKILL” when  both should be recognized as “SPECSKILL” alone)... Does Spark NLP work differently in this regard (i.e., does it mandatorily need IOB format for the entities), compared to other frameworks?
4. In general, should we use “IOB schema” or “BILUO schema” ([source](https://spacy.io/usage/linguistic-features#accessing-ner)) for our named entities, or is it OK to leave our named entities without it (i.e., as seen [here](https://support.prodi.gy/t/ner-format-to-conll/1153))?

I hope your comments. Thank you.

**Comment by maziyarpanahi** [comment, marked as answer]
Hi,

1. Yes! I am not sure about your own CoNLL file, but you don't have any `LABELONE` or `LABELTWO` inside that file. Instead, you have the following labels (Obviously, Spark NLP is not making these labels up so you are either reading the wrong CoNLL index or this is what's actually inside that conll file):

```
U_SPECSKILL
I-SPECSKILL
I-HUMANSKILL
B-SPECSKILL
B-HUMANSKILL
U-HUMANSKILL
```

The number of labels are with `B-` and `I-` plus `O` when they are counted. That's why for CoNLL2003 file you have tested with 4 **entities** (not labels) you have **9 labels**, 8 different labels starting with `B-` and `I-` and `O` which makes them 9 unique labels to learn during the training. You have mistaken the entities with labels. (we teach the model how labels work with beginning and ending schema, then it predicts those back, and then we merge them to entities via NerConverter)

If you just count them you see they are 9 (8+`O`):

```
B-LOC
I-ORG
I-MISC
I-LOC
I-PER
B-MISC
B-ORG
B-PER
O
```

The answer to the rest of your questions, `NerDLApproach` in Spark NLP accepts `IOB` or `IOB2` schemas. However, I've seen people who have datasets without any `B-` or `I-` and only full entities like `PERSON` or `DATE` and the model still learns. I also have seen people with different schema who trained their models. The Schema helps to have more fine-grain NER models with span definition, that's all. (I recommend `IOB` or `IOB2`

Here is a very complete tutorial explaining everything I said in details (this will answer all your questions): https://github.com/JohnSnowLabs/spark-nlp-workshop/blob/master/tutorials/Certification_Trainings/Public/4.NERDL_Training.ipynb

**Reply by ag-din** [reply]
Hi @maziyarpanahi. Thank you very much for your answers, they were clarifying, and the notebook is also helpful.
